src/index.js

```diff
@@ -10,36 +10,123 @@ const assertIsValidText = function(text) {
+    characters === null ||
+    characters === ''
+  ) {
+    length = -1;
+  } else {
+    let charactersArray;
+    if (typeof characters === 'string') {
+      charactersArray = [characters];
+    } else if (Array.isArray(characters) && characters.length) {
+      charactersArray = characters;
+    }
 
-    if (splitAt === -1) {
-      // No whitespace found, we need to truncate the word in that case.
-      const chunk = characters.slice(0, chunkSize).join('');
-      chunks.push(chunk);
-      characters = characters.slice(chunkSize); // eslint-disable-line no-param-reassign
+    if (
+      !Array.isArray(charactersArray) ||
+      !charactersArray.length ||
+      charactersArray === null
+    ) {
+      length = -1;
+    } else if (chunkType === 0) {
+      length = charactersArray
+        .map(character => textEncoder.encode(character).length)
+        .reduce((accumulator, currentValue) => accumulator + currentValue);
+    } else if (chunkType > 0) {
+      const arrayLength = charactersArray
+        .map(character => textEncoder.encode(character).length)
+        .reduce(
+          (accumulator, currentValue) =>
+            accumulator + (currentValue > chunkType ? chunkType : currentValue)
+        );
+      const maxLength = charactersArray.length * chunkType;
+      length = maxLength > arrayLength ? arrayLength : maxLength;
     } else {
-      const chunk = characters.slice(0, splitAt).join('');
-      chunks.push(chunk);
-      characters = characters.slice(splitAt + 1); // eslint-disable-line no-param-reassign
+      length = charactersArray.length;
     }
   }
+  return length;
+};
 
-  chunks.push(characters.join(''));
+const chunkIndexOf = function(characters, chunkSize, chunkType, textEncoder) {
+  let splitAt = characters.lastIndexOf(' ', chunkSize);
+  if (splitAt > -2 && splitAt < 1) {
+    splitAt = chunkSize;
+  }
+  if (splitAt > characters.length || chunkSize >= characters.length) {
+    splitAt = characters.length;
+  }
+  while (
+    splitAt > 0 &&
+    chunkSize <
+      chunkLength(characters.slice(0, splitAt), chunkType, textEncoder)
+  ) {
+    splitAt = splitAt - 1;
+  }
+  if ((splitAt > -2 && splitAt < 1) || characters[splitAt] === ' ') {
+    splitAt = splitAt + 1;
+  }
+  if (
+    splitAt > characters.length ||
+    splitAt < 0 ||
+    (splitAt === 0 && characters.length === 1)
+  ) {
+    splitAt = characters.length;
+  }
+  return splitAt;
+};
 
+export default function(text, chunkSize, chunkType) {
+  assertIsValidText(text);
+  const chunkSizeInt = Number.parseInt(chunkSize, 10);
+  assertIsValidChunkSize(chunkSizeInt);
+  const chunkTypeParseInt = Number.parseInt(chunkType, 10);
+  const chunkTypeParseIntNaN = Number.isNaN(chunkTypeParseInt);
+  assertIsValidChunkType(chunkType, chunkTypeParseIntNaN, chunkTypeParseInt);
+  const chunkTypeInt = chunkTypeParseIntNaN ? -1 : chunkTypeParseInt;
+  const chunks = [];
+  let characters = runes(text);
+  const textEncoder = new TextEncoder();
```

Comment:
important to note that adding this is a breaking change, since it changes browser support without a polyfill

Reply:
😬 understood, hadn't thought of that. I'd figured babel would polyfill it, looking in `dist/`, I see that's not the case.

What do you recommend?

There is the **polyfill in the mdn** link:

https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder

Or the **polyfill mdn suggests for non-utf8**:

https://github.com/inexorabletash/text-encoding

**^ this one** also appears to have an **npm package**, not sure about the mdn polyfill:

https://www.npmjs.com/package/text-encoding

I used **`fast-text-encoding`** for `jest` in dev, but I didn't look at it too closely past for just dev purposes:

https://www.npmjs.com/package/fast-text-encoding

Reply:
I can also backtrack on using TextEncoder, but it seemed like the best of the options when it is available. Whatever works best!

Reply:
Or can call it a breaking change and move on, but I'm unsure your preferences there.

Reply:
Is there some way to make it optional, only using the TextEncoder, and falling back on a more rough algorithm if it's not given, e.g. `chunk(..., TextEncoder)` (which would allow a polyfill to be passed by the user), where the default one is either the global TextEncoder, or a small alternative with the same API

Reply:
(working on this)

Reply:
i've got all this working locally, i have to copy the changes to both branches and still need to fix up some tests. busy with work, hoping this or next weekend i'll have the time